File: client/src/app/services/match-room/match-room.service.ts
import { Injectable } from '@angular/core';
//import { MatDialog } from '@angular/material/dialog';
import { Router } from '@angular/router';
import { ChatChannel } from '@app/constants/chat-channels';
import { MatchContext } from '@app/constants/states';
import { Player } from '@app/interfaces/player';
import { Question } from '@app/interfaces/question';
import { VotingData } from '@app/interfaces/voting-data';
import { ChatService } from '@app/services/chat/chat.service';
import { MatchContextService } from '@app/services/match-context/match-context.service';
import { NotificationService } from '@app/services/notification/notification.service';
import { SocketHandlerService } from '@app/services/socket-handler/socket-handler.service';
import { AnswerEvents } from '@common/events/answer.events';
import { ChatEvents } from '@common/events/chat.events';
import { MatchEvents } from '@common/events/match.events';
import { PartyConfig } from '@common/interfaces/party-config';
import { UserInfo } from '@common/interfaces/user-info';
import { translate } from '@jsverse/transloco';
@Injectable({
    providedIn: 'root',
})
export class MatchRoomService {
    players: Player[];
    isMatchStarted: boolean;
    isResults: boolean;
    isWaitOver: boolean;
    isBanned: boolean;
    isPlaying: boolean;
    gameTitle: string;
    gameDuration: number;
    currentQuestion: Question;
    isHostPlaying: boolean;
    isCooldown: boolean;
    isQuitting: boolean;
    isCheaterMode: boolean;
    cheaterPlayer: Player;
    votesData: VotingData;
    totalVotes: VotingData[] = [{ username: '', numberOfVotes: 0, usersWhoVoted: [] }];
    votesResults: { [username: string]: number };
    partyConfig: PartyConfig;
    votingUsers: string[] = [];
    userVoted: string;
    currentAnswers: string[] = [];
    startedVote: boolean = false;

    private hostId: string;
    private matchRoomCode: string;
    private username: string;
    private userId: string;
    private hasEnteredRoom: boolean;

    // Allow more constructor parameters
    // eslint-disable-next-line max-params
    constructor(
        public socketService: SocketHandlerService,
        private readonly router: Router,
        private readonly notificationService: NotificationService,
        private readonly matchContextService: MatchContextService,
        private chatService: ChatService,
    ) //  private readonly dialog: MatDialog,
    {
        this.hasEnteredRoom = false;
    }

    get socketId() {
        return this.socketService.socket.id ? this.socketService.socket.id : '';
    }

    getRoomCode() {
        return this.matchRoomCode;
    }

    getUsername() {
        return this.username;
    }

    getUserId() {
        return this.userId;
    }

    getHostId() {
        return this.hostId;
    }

    connect() {
        if (!this.hasEnteredRoom) {
            this.hasEnteredRoom = true;
            this.chatService.channel = ChatChannel.ROOM;
            this.resetMatchValues();
            this.onRedirectAfterDisconnection();
            this.onFetchPlayersData();
            this.onMatchStarted();
            this.onMatchCheaterModeStarted();
            this.onSelectedCheater();
            this.onBeginQuiz();
            this.onNextQuestion();
            this.onStartCooldown();
            this.onHostQuit();
            this.onPlayerKick();
            this.handleError();
            this.onPlayerChatStateToggle();
            this.onRouteToResultsPage();
            this.onVotingResults();
            this.onCurrentAnswers();
            this.onVoting();
            this.onUsersWhoVoted();
            //this.goToVoting();
        }
    }

    disconnectFromRoom() {
        this.router.navigateByUrl('/home');
        this.chatService.channel = ChatChannel.GENERAL;
        this.chatService.clearMatchRoomMessages();
        this.hasEnteredRoom = false;
        this.socketService.socket.removeListener(MatchEvents.FetchPlayersData);
        this.socketService.socket.removeListener(MatchEvents.MatchStarting);
        this.socketService.socket.removeListener(MatchEvents.BeginQuiz);
        this.socketService.socket.removeListener(MatchEvents.GoToNextQuestion);
        this.socketService.socket.removeListener(MatchEvents.StartCooldown);
        this.socketService.socket.removeListener(MatchEvents.HostQuitMatch);
        this.socketService.socket.removeListener(MatchEvents.KickPlayer);
        this.socketService.socket.removeListener(MatchEvents.Error);
        this.socketService.socket.removeListener(MatchEvents.RouteToResultsPage);
        this.socketService.socket.removeListener(MatchEvents.CurrentAnswers);
        this.socketService.socket.removeListener(MatchEvents.ShowVotingDialog);
        this.socketService.socket.removeListener(MatchEvents.SendVotingUsers);
        this.socketService.socket.removeListener(MatchEvents.CheaterModeMatchStarting);
        this.socketService.socket.removeListener(MatchEvents.SendBackVotesResults);
        this.socketService.socket.removeListener(AnswerEvents.EndGame);
        this.socketService.send(MatchEvents.Disconnect);
        this.matchContextService.resetContext();
        this.isCheaterMode = false;
        this.votingUsers = [];
        // this.currentAnswers = [];
        // this.socketService.socket.removeListener(MatchEvents.Disconnect);
    }

    sendBackVotesResult(voteData: VotingData) {
        this.socketService.send(MatchEvents.SendVotesResults, voteData);
    }

    createRoom(
        gameId: string,
        hostId: string,
        hostUsername: string,
        isClassicMode: boolean = true,
        partyConfig: PartyConfig = { isFriendsOnly: false, isEntryFeeRequired: false, isCheaterMode: false, canPlayCheaterMode: false },
    ) {
        this.socketService.send(MatchEvents.CreateRoom, { gameId, hostId, isClassicMode, partyConfig }, (res: { code: string }) => {
            if (res) {
                this.matchRoomCode = res.code;
                this.username = hostUsername;
                this.hostId = hostId;
                this.userId = hostId;
                this.partyConfig = partyConfig;

                this.sendPlayersData(this.matchRoomCode);
                this.router.navigateByUrl('/match-room');
            }
        });
    }
    onSelectedCheater() {
        this.socketService.on(MatchEvents.SendCheater, (data: { player: string }) => {
            //TO DO: ADD ERROR HANDLING
            const cheaterPlayer = this.getPlayerByUsername(data.player);
            if (cheaterPlayer) {
                this.cheaterPlayer = cheaterPlayer;
            }
        });
    }

    getPlayerByUsername(username: string): Player | null {
        const playerToFindByUsername = this.players.find((player) => player.username === username);
        return playerToFindByUsername ? playerToFindByUsername : null;
    }

    onPlayerChatStateToggle() {
        this.socketService.on(ChatEvents.ReturnCurrentChatState, (currentChatState: boolean) => {
            const player = this.getPlayerByUsername(this.username);
            if (player) {
                player.isChatActive = currentChatState;
            }
        });
    }

    joinRoom(roomCode: string, username: string, userId: string) {
        const sentInfo: UserInfo = { roomCode, username, userId };
        this.socketService.send(MatchEvents.JoinRoom, sentInfo, (res: { code: string; userId: string; username: string }) => {
            if (res) {
                this.matchRoomCode = res.code;
                this.username = res.username;
                this.userId = res.userId;
                this.router.navigateByUrl('/match-room');
                this.sendPlayersData(roomCode);
            }
        });
    }

    sendPlayersData(roomCode: string) {
        this.socketService.send(MatchEvents.SendPlayersData, roomCode);
    }

    banUser(userId: string) {
        if (this.userId === this.hostId) {
            const sentInfo: UserInfo = { roomCode: this.matchRoomCode, userId };
            this.socketService.send(MatchEvents.BanUsername, sentInfo);
        }
    }

    handleError() {
        this.socketService.on(MatchEvents.Error, (errorMessage: string) => {
            this.notificationService.displayErrorMessage(errorMessage);
        });
    }

    startMatch() {
        this.isMatchStarted = true;
        this.socketService.send(MatchEvents.StartMatch, this.matchRoomCode);
    }

    startMatchCheaterMode() {
        this.isCheaterMode = true;
        this.isMatchStarted = true; // we can keep the same.
        this.socketService.send(MatchEvents.StartMatchCheaterMode, this.matchRoomCode);
    }

    onMatchCheaterModeStarted() {
        this.socketService.on(MatchEvents.CheaterModeMatchStarting, (data: { start: boolean; gameTitle: string }) => {
            this.isCheaterMode = true;
            if (data.start) {
                this.isMatchStarted = data.start;
            }
            if (data.gameTitle) {
                this.gameTitle = data.gameTitle;
            }
        });
    }

    onMatchStarted() {
        this.socketService.on(MatchEvents.MatchStarting, (data: { start: boolean; gameTitle: string }) => {
            if (data.start) {
                this.isMatchStarted = data.start;
            }
            if (data.gameTitle) {
                this.gameTitle = data.gameTitle;
            }
        });
    }

    onBeginQuiz() {
        console.log('user', this.userId);
        console.log(this.cheaterPlayer?.id);
        this.socketService.on(MatchEvents.BeginQuiz, (data: { firstQuestion: Question; gameDuration: number; isClassicMode: boolean }) => {
            this.isWaitOver = true;
            this.currentQuestion = data.firstQuestion;
            this.gameDuration = data.gameDuration;
            const { firstQuestion, gameDuration } = data;
            this.router.navigate(['/play-match'], { state: { question: firstQuestion, duration: gameDuration } });
        });
    }

    goToNextQuestion() {
        this.socketService.send(MatchEvents.GoToNextQuestion, this.matchRoomCode);
    }

    voteOnCheater() {
        this.socketService.send(MatchEvents.VoteOnCheater, this.matchRoomCode);
    }

    onVoting() {
        this.socketService.on(MatchEvents.ShowVotingDialog, () => {
            this.router.navigateByUrl('/vote');
        });
    }

    onUsersWhoVoted() {
        this.socketService.on(MatchEvents.SendVotingUsers, (user: string) => {
            console.log('Voted', this.votingUsers);
            
            this.userVoted = user;
            console.log("user", this.userVoted)
            this.votingUsers.push(user);
        });
    }
    onVotingResults() {
        this.socketService.on(MatchEvents.SendBackVotesResults, (data: { [username: string]: number }) => {
            this.votesResults = data;
            console.log(data);
        });
    }

    onStartCooldown() {
        this.socketService.on(MatchEvents.StartCooldown, () => {
            this.isCooldown = true;
            const context = this.matchContextService.getContext();
            if (this.isCooldown && context !== MatchContext.TestPage && context !== MatchContext.RandomMode) {
                this.currentQuestion.text = translate('feedback-messages.prepare');
            }
        });
    }

    onNextQuestion() {
        this.socketService.on(MatchEvents.GoToNextQuestion, (question: Question) => {
            this.isCooldown = false;
            this.currentQuestion = question;
        });
    }

    onFetchPlayersData() {
        this.socketService.on(MatchEvents.FetchPlayersData, (res: string) => {
            this.players = JSON.parse(res);
        });
    }

    onHostQuit() {
        this.socketService.on(MatchEvents.HostQuitMatch, () => {
            this.isHostPlaying = false;
            this.disconnectFromRoom();
        });
    }

    onRedirectAfterDisconnection() {
        this.socketService.on(MatchEvents.Disconnect, () => {
            this.router.navigateByUrl('/home');
            this.resetMatchValues();
        });
    }

    resetMatchValues() {
        this.matchRoomCode = '';
        this.username = '';
        this.players = [];
        this.isResults = false;
        this.isWaitOver = false;
        this.isPlaying = false;
        this.isCooldown = false;
        this.currentAnswers = [];
        this.votesResults = {};
        this.votesData = { username: '', numberOfVotes: 0, usersWhoVoted: [] };
        this.votingUsers = [];
        this.resetCheaterPlayerValue();
        this.chatService.clearMatchRoomMessages();
    }

    routeToResultsPage() {
        this.socketService.send(MatchEvents.RouteToResultsPage, this.matchRoomCode);
    }

    resetCheaterPlayerValue() {
        if (this.cheaterPlayer?.username != '') {
            this.cheaterPlayer = {
                username: '',
                id: '',
                photoUrl: '',
                score: 0,
                bonusCount: 0,
                isPlaying: false,
                isChatActive: false,
                state: '',
            };
        }
        this.startedVote = false;
    }

    onRouteToResultsPage() {
        this.socketService.on(MatchEvents.RouteToResultsPage, () => {
            this.isResults = true;
            this.router.navigateByUrl('/results');
        });
    }

    onPlayerKick() {
        this.socketService.on(MatchEvents.KickPlayer, () => {
            this.isBanned = true;
            this.disconnectFromRoom();
        });
    }

    toggleLock() {
        this.socketService.send(MatchEvents.ToggleLock, this.matchRoomCode);
    }

    onCurrentAnswers() {
        this.socketService.on(MatchEvents.CurrentAnswers, (answer: string[]) => {
            if (this.userId === this.hostId || this.userId === this.cheaterPlayer?.id) {
                console.log('Current answers:', answer);
                this.currentAnswers = answer;
            }
           // return;
        });
    }
}

File: client-mobile/app/src/main/java/com/example/polyquiz/match/domain/MatchRoomService.kt
package com.example.polyquiz.match.domain

import StringValue
import android.annotation.SuppressLint
import android.content.Context
import com.example.polyquiz.constants.MatchContext
import com.example.polyquiz.constants.MatchEvents
import com.example.polyquiz.constants.MatchStatus
import com.example.polyquiz.chat.domain.Message
import com.example.vanillaprototype.socket.SocketHandler
import com.google.gson.Gson
import io.socket.client.Ack
import org.json.JSONObject
import android.util.Log
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.res.stringResource
import com.example.polyquiz.R
import com.example.polyquiz.SnackbarController
import com.example.polyquiz.SnackbarEvent
import com.example.polyquiz.chat.domain.ChatService
import com.example.polyquiz.constants.AnswerEvents
import com.example.polyquiz.constants.ChatEvents
import com.example.polyquiz.constants.LongAnswerInfo
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import com.example.polyquiz.constants.Route
import com.example.polyquiz.constants.VotingData
import com.example.polyquiz.pages.presentation.VotingPage
import com.example.polyquiz.elo.domain.EloService
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import com.example.polyquiz.match.domain.AnswerService.gradeAnswers
import com.example.polyquiz.match.domain.AnswerService.playersAnswers
import com.google.gson.reflect.TypeToken
import kotlinx.coroutines.flow.asStateFlow
import org.json.JSONArray
import org.json.JSONException
import java.util.Arrays

@SuppressLint("StaticFieldLeak")
object MatchRoomService {
    var players by mutableStateOf<List<Player>>(emptyList())
    var messages by mutableStateOf<List<Message>>(emptyList())
    var isMatchStarted by mutableStateOf(false)
    var isResults by mutableStateOf(false)
    var isWaitOver by mutableStateOf(false)
    var isBanned by mutableStateOf(false)
    var timeToGoToWaitPage by mutableStateOf(false)
    var isPlaying by mutableStateOf(false)
    var isTimeToNavigate by mutableStateOf(false)
    var isTimeToNavigateToResults by mutableStateOf(false)
    var hasBeenKickedOut by mutableStateOf(false)
    var isLocked by mutableStateOf(false)
    var canPlayCheaterMode by mutableStateOf(false)
    var gameTitle: String = ""
    var gameDuration: Int = 0
    var currentAnswers: MutableList<String> = mutableListOf()
    var partyConfig by mutableStateOf(PartyConfig(false, false, 0, false, false))
    var currentQuestion by mutableStateOf<Question?>(null)
    var isHostPlaying by mutableStateOf(true)
    var isCooldown by mutableStateOf(false)
    var isQuitting by mutableStateOf(false)
    var username by mutableStateOf("")
    var userId by mutableStateOf("")
    var hostId by mutableStateOf("")
    var errorMsg by mutableStateOf("")
    var navigateToVotingPage by mutableStateOf(false)
    var votingUsers = MutableStateFlow<List<String>>(emptyList())
    var playerVoted  by mutableStateOf(false)

    var userVoted by mutableStateOf("")
    var votesResults = mutableMapOf<String, Int>()
    var startedVote = false

    var cheaterPlayer by mutableStateOf(
        Player(
            "", "", "", 0, 0,
            false, false, ""
        )
    )

    var isCheaterMode by mutableStateOf(false)
    var votesData by mutableStateOf(VotingData("", 0, mutableListOf("")))
    var totalVotes: MutableList<VotingData> = mutableListOf()
    private var hasEnteredRoom = false
    private val _matchRoomCode = MutableStateFlow("")
    val matchRoomCode: StateFlow<String> get() = _matchRoomCode

    val socket = SocketHandler.getSocket()

    val socketId: String
        get() = socket.id() ?: ""

    fun getRoomCode(): String = _matchRoomCode.value
    fun retrieveUsername(): String = username

    fun connect(context: Context) {
        if (!hasEnteredRoom) {
            hasEnteredRoom = true
            resetMatchValues()
            onRedirectAfterDisconnection()
            onFetchPlayersData()
            onMatchStarted()
            onBeginQuiz()
            onNextQuestion()
            onStartCooldown()
            onHostQuit()
            onPlayerKick()
            handleError(context)
//            onPlayerChatStateToggle()
            onRouteToResultsPage()
        }
    }

    fun disconnectFromRoom() {
        hasEnteredRoom = false
        socket.off(MatchEvents.FETCH_PLAYERS_DATA.value)
        socket.off(MatchEvents.MATCH_STARTING.value)
        socket.off(MatchEvents.BEGIN_QUIZ.value)
        socket.off(MatchEvents.GO_TO_NEXT_QUESTION.value)
        socket.off(MatchEvents.START_COOLDOWN.value)
        socket.off(MatchEvents.HOST_QUIT_MATCH.value)
        socket.off(MatchEvents.KICK_PLAYER.value)
        socket.off(MatchEvents.ERROR.value)
        socket.off(MatchEvents.ROUTE_TO_RESULTS_PAGE.value)
        socket.off(ChatEvents.NEW_MESSAGE.value)
        socket.off(ChatEvents.SENT_ROOM_EMOJI.value)
        socket.off(MatchEvents.SEND_BACK_VOTES_RESULTS.value)
        socket.off(MatchEvents.CURRENT_ANSWERS.value)
        socket.off(MatchEvents.CHEATER_MODE_MATCH_STARTING.value)
        socket.off(MatchEvents.SHOW_VOTING_DIALOG.value)
        socket.off(MatchEvents.SEND_VOTING_USERS.value)
        socket.off(AnswerEvents.END_GAME.value)
        ChatService.deleteRoomMessages()
        socket.emit(MatchEvents.DISCONNECT.value)
        MatchContextService.resetContext()
        hostId = ""
        timeToGoToWaitPage = false
        hasBeenKickedOut = true
        navigateToVotingPage = false
        isCheaterMode = false
        resetMatchValues()
        Log.d("Disconnect from room WaitPage", "Called disconnectFromRoom, hostId=$hostId")
        isTimeToNavigateToResults = false
    }

    fun sendBackVotesResult(voteData: JSONObject) {
        socket.emit(MatchEvents.SEND_VOTES_RESULTS.value, voteData)
    }

    fun onUsersWhoVoted() {
        socket.on(MatchEvents.SEND_VOTING_USERS.value) { args ->
            if (args.isNotEmpty()) {
                val firstArg = args[0]
                if (firstArg is JSONArray) {
                    if (firstArg.length() > 0) {
                        val user = firstArg.getString(0)
                        userVoted = user
                    } else {
                        Log.d("onUsersWhoVoted","JSONArray object is empty.")
                    }
                }
            }
        }
    }


    fun startMatchCheaterMode() {
        isCheaterMode = true;
        isMatchStarted = true;
        socket.emit(MatchEvents.START_MATCH_CHEATER_MODE.value, matchRoomCode.value)

    }

    fun onSelectedCheater() {
        socket.on(MatchEvents.SEND_CHEATER.value) { args ->
            if (args.isNotEmpty()) {
                val jsonString = (args[0] as? JSONObject)?.toString() ?: ""
                val jsonObject = JSONObject(jsonString)
                val playerName = jsonObject.optString("player")
                val player = getPlayerByUsername(playerName);
                cheaterPlayer = player!!;
            }
        }
    }

    fun onMatchCheaterModeStarted() {
        socket.on(MatchEvents.CHEATER_MODE_MATCH_STARTING.value) { args ->
            isCheaterMode = true;
            if (args.isNotEmpty()) {
                val data = args[0] as JSONObject
                if (data.optBoolean("start", false)) {
                    isMatchStarted = true
                }
                if (data.has("gameTitle")) {
                    gameTitle = data.getString("gameTitle")
                }
            }
        }
    }

    fun onVoting() {
        socket.on(MatchEvents.SHOW_VOTING_DIALOG.value) {
            navigateToVotingPage = true
        }
    }


    fun voteOnCheater() {
        socket.emit(MatchEvents.VOTE_ON_CHEATER.value, matchRoomCode.value)
    }

    fun onCurrentAnswers() {
        socket.on(MatchEvents.CURRENT_ANSWERS.value) { args ->
            if (userId == hostId || userId == cheaterPlayer?.id) {
                if (args.isNotEmpty() && args[0] != null) {
                    currentAnswers = Gson().fromJson(
                        args[0].toString(),
                        object : TypeToken<List<String>>() {}.type
                    )
                }
            }
        }
    }

    fun onVotingResults() {
        socket.on(MatchEvents.SEND_BACK_VOTES_RESULTS.value) { args ->
            if (args.isNotEmpty()) {
                val data = args[0] as JSONObject
                val gson = Gson()
                val votingData = gson.fromJson(data.toString(), VotingData::class.java)
                votesData = votingData
                totalVotes.add(votingData)

                val votesMap = mutableMapOf<String, Int>()
                val iterator = data.keys()
                while (iterator.hasNext()) {
                    val key = iterator.next()
                    val value = data.opt(key.toString())
                    if (value is Int) {
                        votesMap[key.toString()] = value
                    } else {
                        votesMap[key.toString()] = 0
                    }
                }
                votesResults = votesMap
            }
        }
    }

    fun createRoom(
        gameId: String,
        hostId: String,
        hostUsername: String,
        isClassicMode: Boolean = true,
        partyConfigs: PartyConfig = PartyConfig(false, false, isCheaterMode = false)
    ) {
        val partyConfisObject = JSONObject().apply {
            put("isFriendsOnly", partyConfigs.isFriendsOnly)
            put("isEntryFeeRequired", partyConfigs.isEntryFeeRequired)
            put("entryFeeAmount", partyConfigs.entryFeeAmount)
            put("isCheaterMode", partyConfigs.isCheaterMode)
        }
        val data = JSONObject().apply {
            put("gameId", gameId)
            put("hostId", hostId)
            put("partyConfig", partyConfisObject)
            put("isClassicMode", isClassicMode)
        }
        println("data : $data")

        socket.emit(MatchEvents.CREATE_ROOM.value, data, Ack { args ->
            if (args.isNotEmpty()) {
                val response = args[0] as JSONObject
                _matchRoomCode.value = response.getString("code")
                username = hostUsername
                userId = hostId
                partyConfig = partyConfigs
                this.hostId = hostId
                sendPlayersData(_matchRoomCode.value)
            }
        })
    }

    fun getPlayerByUsername(username: String): Player? =
        players.find { it.username == username }

    fun sendPlayersData(roomCode: String) {
        socket.emit(MatchEvents.SEND_PLAYERS_DATA.value, roomCode)
    }

    fun joinRoom(roomCode: String, username: String, userId: String) {
        Log.d("Join room", "Hostid : $hostId")
        val sentInfo = JSONObject().apply {
            put("roomCode", roomCode)
            put("username", username)
            put("userId", userId)
        }

        socket.emit(MatchEvents.JOIN_ROOM.value, sentInfo, Ack { args ->
            if (args.isNotEmpty()) {
                ChatService.handleRoomMessage()
                val response = args[0] as JSONObject
                _matchRoomCode.value = response.getString("code")
                this.username = response.getString("username")
                this.userId = response.getString("userId")
                sendPlayersData(roomCode)
                timeToGoToWaitPage = true
            }
        })
    }


    fun banUsername(userId: String) {
        if (this.userId == this.hostId) {
            val sentInfo = JSONObject().apply {
                put("roomCode", matchRoomCode)
                put("userId", userId)
            }
            socket.emit(MatchEvents.BAN_USERNAME.value, sentInfo)
        }
    }

    fun handleError(context: Context) {
        socket.on(MatchEvents.ERROR.value) { args ->
            if (args.isNotEmpty()) {
                val errors: List<String> = try {
                    Gson().fromJson(args[0].toString(), Array<String>::class.java).toList()
                } catch (e: Exception) {
                    listOf(args[0].toString())
                }

                val displayText = errors.joinToString(separator = "\n") { errorKey: String ->
                    val cleanKey = errorKey.trim()
                    StringValue.dynamicLookup(context, cleanKey).asString(context)
                }

                val event = SnackbarEvent(
                    message = StringValue.DynamicString(displayText)
                )

                CoroutineScope(Dispatchers.Main).launch {
                    SnackbarController.sendEvent(event)
                }
            }
        }
    }

    fun startMatch() {
        isMatchStarted = true
        socket.emit(MatchEvents.START_MATCH.value, matchRoomCode.value)
    }

    fun onMatchStarted() {
        socket.on(MatchEvents.MATCH_STARTING.value) { args ->
            if (args.isNotEmpty()) {
                val data = args[0] as JSONObject
                if (data.optBoolean("start", false)) {
                    isMatchStarted = true
                }
                if (data.has("gameTitle")) {
                    gameTitle = data.getString("gameTitle")
                }
            }
        }
    }

    fun onBeginQuiz() {
        socket.on(MatchEvents.BEGIN_QUIZ.value) { args ->
            if (args.isNotEmpty()) {
                val data = args[0] as JSONObject
                isWaitOver = true
                val gson = Gson()
                val firstQuestion = gson.fromJson(
                    data.getJSONObject("firstQuestion").toString(),
                    Question::class.java
                )
                currentQuestion = firstQuestion
                gameDuration = data.getInt("gameDuration")
                isTimeToNavigate = true
            }
        }
    }

    fun goToNextQuestion() {
        socket.emit(MatchEvents.GO_TO_NEXT_QUESTION.value, matchRoomCode.value)
    }

    fun onStartCooldown() {
        socket.on(MatchEvents.START_COOLDOWN.value) { _ ->
            isCooldown = true
            val context = MatchContextService.getContext()
            if (isCooldown && context != MatchContext.TESTPAGE && context != MatchContext.RANDOMMODE) {
                currentQuestion?.text = StringValue.StringResource(R.string.prepare).toString()
            }
        }
    }

    fun onNextQuestion() {
        socket.on(MatchEvents.GO_TO_NEXT_QUESTION.value) { args ->
            if (args.isNotEmpty()) {
                isCooldown = false

                val jsonString = (args[0] as? JSONObject)?.toString() ?: ""

                if (jsonString.isNotEmpty()) {
                    val gson = Gson()
                    currentQuestion = gson.fromJson(jsonString, Question::class.java)
                }
            }
        }
    }

    fun onFetchPlayersData() {
        socket.on(MatchEvents.FETCH_PLAYERS_DATA.value) { args ->
            if (args.isNotEmpty()) {
                val playersJson = args[0] as? String
                playersJson?.let {
                    players = Gson().fromJson(it, Array<Player>::class.java).toList()
                }
            }
        }
    }

    fun onHostQuit() {
        socket.on(MatchEvents.HOST_QUIT_MATCH.value) { _ ->
            isHostPlaying = false
            disconnectFromRoom()
        }
    }

    fun onRedirectAfterDisconnection() {
        socket.on(MatchEvents.DISCONNECT.value) { _ ->
            resetMatchValues()
        }
    }

    fun resetMatchValues() {
        _matchRoomCode.value = ""
        username = ""
        players = emptyList()
        messages = emptyList()
        isResults = false
        isWaitOver = false
        isPlaying = false
        isCooldown = false
        votingUsers = MutableStateFlow(emptyList())
        userVoted = ""
        votesResults = mutableMapOf<String, Int>()
        startedVote = false
        if (cheaterPlayer.username != "") {
            cheaterPlayer = Player("", "", "", 0, 0, false, false, "")
        }
        playerVoted = false
    }

    fun routeToResultsPage() {
        socket.emit(MatchEvents.ROUTE_TO_RESULTS_PAGE.value, matchRoomCode.value)
    }

    fun onRouteToResultsPage() {
        socket.on(MatchEvents.ROUTE_TO_RESULTS_PAGE.value) { _ ->
            println("Navigating to results page")
            isResults = true
            isTimeToNavigateToResults = true
        }
    }


    fun onPlayerKick() {
        socket.on(MatchEvents.KICK_PLAYER.value) { _ ->
            isBanned = true
            disconnectFromRoom()
        }
    }

    fun toggleLock() {
        socket.emit(MatchEvents.TOGGLE_LOCK.value, matchRoomCode.value)
        isLocked = !isLocked
    }
}
